tests/http/server/_base.py

```diff
@@ -98,3 +111,413 @@ def _test_disconnect(
+            make_request: A function that initiates the request and returns a
+                `FakeChannel`.
+            test_name: The name of the test, which will be logged.
+            expected_code: The expected status code for the final request that runs to
+                completion. Defaults to `200`.
+
+        Returns:
+            The JSON response of the final request that runs to completion.
+        """
+        # To process a request, a coroutine run is created for the async method handling
+        # the request. That method may then start other coroutine runs, wrapped in
+        # `Deferred`s.
+        #
+        # We would like to trigger a cancellation at the first `await`, re-run the
+        # request and cancel at the second `await`, and so on. By patching
+        # `Deferred.__next__`, we can intercept `await`s, track which ones we have or
+        # have not seen, and force them to block when they wouldn't have.
+
+        # The set of previously seen `await`s.
+        # Each element is a stringified stack trace.
+        seen_awaits: Set[Tuple[str, ...]] = set()
+
+        self._log_for_request(
+            0, f"Running _test_cancellation_at_every_await for {test_name}..."
+        )
+
+        for request_number in itertools.count(1):
+            (
+                Deferred___next__,
+                unblock_awaits,
+                get_awaits_seen,
+                has_seen_new_await,
+            ) = self.create_deferred___next___patch(seen_awaits, request_number)
+
+            try:
+                with mock.patch(
+                    "synapse.http.server.respond_with_json", wraps=respond_with_json
+                ) as respond_mock:
+                    with mock.patch.object(
+                        Deferred,
+                        "__next__",
+                        new=Deferred___next__,
+                    ):
+                        # Start the request.
+                        channel = make_request()
+                        request = channel.request
+
+                        if request_number == 0:
+                            self.assertFalse(
+                                respond_mock.called,
+                                "Request finished before we could disconnect - "
+                                "was `await_result=False` passed to `make_request`?",
+                            )
+                        else:
+                            # Requests after the first may be lucky enough to hit caches
+                            # all the way through and never have to block.
+                            pass
+
+                        # Run the request until we see a new `await` which we have not
+                        # yet cancelled at, or it completes.
+                        while not respond_mock.called and not has_seen_new_await():
+                            previous_awaits_seen = get_awaits_seen()
+
+                            reactor.pump([0.0])
```

Comment:
I'd inline this (and the other calls to `reactor.pump`) for clarity:

```suggestion
                            reactor.advance(0.0)
```